pkg/operator/encryption/shared_encryption.go

```diff
@@ -0,0 +1,867 @@
+const encryptionSecretMigrationInterval = time.Hour * 24 * 7 // one week
+
+// In the data field of the secret API object, this (map) key is used to hold the actual encryption key
+// (i.e. for AES-CBC mode the value associated with this map key is 32 bytes of random noise).
+const encryptionSecretKeyData = "encryption.apiserver.operator.openshift.io-key"
+
+// These annotations try to scare anyone away from editing the encryption secrets.  It is trivial for
+// an external actor to break the invariants of the state machine and render the cluster unrecoverable.
+const (
+	kubernetesDescriptionKey        = "kubernetes.io/description"
+	kubernetesDescriptionScaryValue = `WARNING: DO NOT EDIT.
+Altering of the encryption secrets will render you cluster inaccessible.
+Catastrophic data loss can occur from the most minor changes.`
+)
+
+// encryptionConfSecret is the name of the final encryption config secret that is revisioned per apiserver rollout.
+// it also serves as the (map) key that is used to store the raw bytes of the final encryption config.
+const encryptionConfSecret = "encryption-config"
+
+// revisionLabel is used to find the current revision for a given API server.
+const revisionLabel = "revision"
+
+// these can only be used to encode and decode EncryptionConfiguration objects
+var (
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+)
+
+func init() {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	utilruntime.Must(apiserverconfigv1.AddToScheme(scheme))
+	utilruntime.Must(apiserverconfig.AddToScheme(scheme))
+	encoder = codecs.LegacyCodec(apiserverconfigv1.SchemeGroupVersion)
+	decoder = codecs.UniversalDecoder(apiserverconfigv1.SchemeGroupVersion)
+}
+
+// groupResourcesState represents the secrets (i.e. encryption keys) associated with each group resource.
+// see getDesiredEncryptionState for how this map is built.  it is first fed the current state based on the on
+// disk configuration.  the actual state of the secrets in the kube API is then layered on top to determine the
+// overall desired configuration (which is the same as the current state when the system is at steady state).
+type groupResourcesState map[schema.GroupResource]keysState
+type keysState struct {
+	targetNamespace string
+
+	readSecrets []*corev1.Secret
+	writeSecret *corev1.Secret
+}
+
+func (k keysState) readKeys() []keyAndMode {
+	ret := make([]keyAndMode, 0, len(k.readSecrets))
+	for _, readKey := range k.readSecrets {
+		readKeyAndMode, _, ok := secretToKeyAndMode(readKey, k.targetNamespace)
+		if !ok {
+			klog.Infof("failed to convert read secret %s to key", readKey.Name)
+			continue
+		}
+		ret = append(ret, readKeyAndMode)
+	}
+	return ret
+}
+
+func (k keysState) writeKey() keyAndMode {
+	if k.writeSecret == nil {
+		return keyAndMode{}
+	}
+
+	writeKeyAndMode, _, ok := secretToKeyAndMode(k.writeSecret, k.targetNamespace)
+	if !ok {
+		klog.Infof("failed to convert write secret %s to key", k.writeSecret.Name)
```

Comment:
this smells. A getter should never log, and always do what it says: return the key. If that's not always possible, add error.